File: backend/services/stigmer-server/pkg/domain/agentexecution/approval/compute_from_events.go
package approval

import (
	agentexecutionv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/agentexecution/v1"
)

// ComputePendingApprovalsFromEvents projects pending approvals from the
// append-only approval-event stream: every REQUESTED event whose approval has
// not been resolved by a later decision event.
//
// This is the read side of the Phase-1 shadow projection. For a given execution
// it must produce the same set as ComputePendingApprovals (the message scan);
// ProjectPendingApprovals asserts that equality in production and the fixture
// corpus asserts it in CI. Correlation is by approval_request_id (in Phase 1,
// the tool_call_id).
func ComputePendingApprovalsFromEvents(
	stream *agentexecutionv1.ApprovalEventStream,
) []*agentexecutionv1.PendingApproval {
	if stream == nil {
		return nil
	}

	decided := make(map[string]struct{})
	for _, ev := range stream.GetEvents() {
		if isDecisionEvent(ev.GetEventType()) {
			decided[ev.GetApprovalRequestId()] = struct{}{}
		}
	}

	var result []*agentexecutionv1.PendingApproval
	for _, ev := range stream.GetEvents() {
		if ev.GetEventType() != agentexecutionv1.ApprovalEventType_APPROVAL_EVENT_TYPE_REQUESTED {
			continue
		}
		if _, resolved := decided[ev.GetApprovalRequestId()]; resolved {
			continue
		}
		if req := ev.GetRequested(); req != nil {
			result = append(result, pendingApprovalFromRequest(req))
		}
	}
	return result
}

func isDecisionEvent(t agentexecutionv1.ApprovalEventType) bool {
	switch t {
	case agentexecutionv1.ApprovalEventType_APPROVAL_EVENT_TYPE_APPROVED,
		agentexecutionv1.ApprovalEventType_APPROVAL_EVENT_TYPE_REJECTED,
		agentexecutionv1.ApprovalEventType_APPROVAL_EVENT_TYPE_SKIPPED:
		return true
	default:
		return false
	}
}

// pendingApprovalFromRequest reconstructs a PendingApproval from a REQUESTED
// event's payload. ApprovalRequest intentionally carries the same display fields
// as PendingApproval so this projection needs no join back to the ToolCall.
func pendingApprovalFromRequest(req *agentexecutionv1.ApprovalRequest) *agentexecutionv1.PendingApproval {
	return &agentexecutionv1.PendingApproval{
		ToolCallId:      req.GetToolCallId(),
		ToolName:        req.GetToolName(),
		Message:         req.GetMessage(),
		ArgsPreview:     req.GetArgsPreview(),
		RequestedAt:     req.GetRequestedAt(),
		FromSubAgent:    req.GetFromSubAgent(),
		SubAgentName:    req.GetSubAgentName(),
		SubAgentSubject:      req.GetSubAgentSubject(),
		McpServerSlug:        req.GetMcpServerSlug(),
		ToolKind:             req.GetToolKind(),
		ApprovalPolicySource: req.GetApprovalPolicySource(),
		FileChanges:          req.GetFileChanges(),
	}
}

File: backend/services/stigmer-server-ts/src/domain/agentexecution/approval/compute-from-events.ts
/**
 * The pending-approvals EVENT-STREAM projection — ports
 * approval/compute_from_events.go: every REQUESTED event whose approval
 * has not been resolved by a later terminal event — a user decision
 * (APPROVED/REJECTED/SKIPPED) or a system RETRACTED.
 *
 * This is the AUTHORITATIVE projection: projectPendingApprovals returns
 * its result. For a given execution it must produce the same set as the
 * retained message-scan cross-check; project.ts asserts that equality in
 * production and the shared HITL corpus asserts it in CI. The RETRACTED
 * resolution is what lets the two agree for the in-flight orphan exits
 * the scan handles structurally — see reconcileRetractions (author.ts).
 * Terminal-execution gate-exits are handled one level up, by the
 * phase-aware seam, not here. Correlation is by approval_request_id
 * (today, the tool_call_id).
 */
import { create } from "@bufbuild/protobuf";

import type {
  ApprovalEventStream,
  ApprovalRequest,
  PendingApproval,
} from "@stigmer/protos/ai/stigmer/agentic/agentexecution/v1/approval_pb";
import { PendingApprovalSchema } from "@stigmer/protos/ai/stigmer/agentic/agentexecution/v1/approval_pb";
import { ApprovalEventType } from "@stigmer/protos/ai/stigmer/agentic/agentexecution/v1/enum_pb";

export function computePendingApprovalsFromEvents(
  stream: ApprovalEventStream | undefined,
): PendingApproval[] {
  if (stream === undefined) {
    return [];
  }

  const resolved = new Set<string>();
  for (const ev of stream.events) {
    if (isResolvingEvent(ev.eventType)) {
      resolved.add(ev.approvalRequestId);
    }
  }

  const result: PendingApproval[] = [];
  for (const ev of stream.events) {
    if (ev.eventType !== ApprovalEventType.REQUESTED) {
      continue;
    }
    if (resolved.has(ev.approvalRequestId)) {
      continue;
    }
    if (ev.payload.case === "requested") {
      result.push(pendingApprovalFromRequest(ev.payload.value));
    }
  }
  return result;
}

/**
 * Whether an event type terminally resolves a REQUESTED: the three user
 * decisions plus the system RETRACTED. A resolved request never appears
 * in the pending projection.
 */
export function isResolvingEvent(t: ApprovalEventType): boolean {
  switch (t) {
    case ApprovalEventType.APPROVED:
    case ApprovalEventType.REJECTED:
    case ApprovalEventType.SKIPPED:
    case ApprovalEventType.RETRACTED:
      return true;
    default:
      return false;
  }
}

/**
 * Reconstructs a PendingApproval from a REQUESTED event's payload.
 * ApprovalRequest intentionally carries the same display fields as
 * PendingApproval so this projection needs no join back to the ToolCall.
 */
function pendingApprovalFromRequest(req: ApprovalRequest): PendingApproval {
  return create(PendingApprovalSchema, {
    toolCallId: req.toolCallId,
    toolName: req.toolName,
    message: req.message,
    argsPreview: req.argsPreview,
    requestedAt: req.requestedAt,
    fromSubAgent: req.fromSubAgent,
    subAgentName: req.subAgentName,
    subAgentSubject: req.subAgentSubject,
    mcpServerSlug: req.mcpServerSlug,
    toolKind: req.toolKind,
    approvalPolicySource: req.approvalPolicySource,
  });
}
